examples/mm/stable_diffusion/train_sd_draftp.py

```diff
@@ -0,0 +1,118 @@
+from nemo_aligner.data.mm import text_webdataset
+from nemo_aligner.data.nlp.builders import build_dataloader
+from nemo_aligner.models.mm.stable_diffusion.image_text_rms import get_reward_model
+from nemo_aligner.models.mm.stable_diffusion.megatron_sd_draftp_model import MegatronSDDRaFTPModel
+from nemo_aligner.utils.distributed import Timer
+from nemo_aligner.utils.train_script_utils import (
+    CustomLoggerWrapper,
+    add_custom_checkpoint_callback,
+    extract_optimizer_scheduler_from_ptl_model,
+    init_distributed,
+    init_peft,
+    init_using_ptl,
+    retrieve_custom_trainer_state_dict,
+)
+
+mp.set_start_method("spawn", force=True)
+
+
+@hydra_runner(config_path="conf", config_name="draftp_sd")
+def main(cfg) -> None:
+
+    logging.info("\n\n************** Experiment configuration ***********")
+    logging.info(f"\n{OmegaConf.to_yaml(cfg)}")
+
+    # TODO: has to be set true for PyTorch 1.12 and later.
+    torch.backends.cuda.matmul.allow_tf32 = True
+    cfg.model.data.train.dataset_path = [cfg.model.data.webdataset.local_root_path for _ in range(cfg.trainer.devices)]
+    trainer = MegatronStableDiffusionTrainerBuilder(cfg).create_trainer()
+    exp_manager(trainer, cfg.exp_manager)
+    logger = CustomLoggerWrapper(trainer.loggers)
+    # Instatiating the model here
+    ptl_model = MegatronSDDRaFTPModel(cfg.model, trainer).to(torch.cuda.current_device())
+    init_peft(ptl_model, cfg.model)
+
+    trainer_restore_path = trainer.ckpt_path
+
+    if trainer_restore_path is not None:
+        custom_trainer_state_dict = retrieve_custom_trainer_state_dict(trainer)
+        consumed_samples = custom_trainer_state_dict["consumed_samples"]
+    else:
+        custom_trainer_state_dict = None
+        consumed_samples = 0
+
+    init_distributed(trainer, ptl_model, cfg.model.get("transformer_engine", False))
+
+    train_dataset, _ = text_webdataset.build_train_valid_datasets(cfg.model.data, consumed_samples=consumed_samples)
+    train_dataset = [d["captions"] for d in list(train_dataset)]
+
+    train_dataloader = build_dataloader(
+        cfg,
+        dataset=train_dataset,
+        consumed_samples=consumed_samples,
+        mbs=cfg.model.micro_batch_size,
+        gbs=cfg.model.global_batch_size,
+        load_gbs=True,
+    )
+
+    init_using_ptl(trainer, ptl_model, train_dataloader, train_dataset)
+
+    optimizer, scheduler = extract_optimizer_scheduler_from_ptl_model(ptl_model)
+
+    ckpt_callback = add_custom_checkpoint_callback(trainer, ptl_model)
+
+    logger.log_hyperparams(OmegaConf.to_container(cfg))
+
+    reward_model = get_reward_model(cfg.rm, mbs=cfg.model.micro_batch_size, gbs=cfg.model.global_batch_size)
+    ptl_model.reward_model = reward_model
+    ptl_model.wandb_logger = logger
+
+    ckpt_callback = add_custom_checkpoint_callback(trainer, ptl_model)
+    timer = Timer(cfg.exp_manager.get("max_time_per_run", "0:12:00:00"))
+
+    draft_p_trainer = SupervisedTrainer(
+        cfg=cfg.model,
```

Comment:
this by convention should be cfg.trainer.sd

Reply:
see how we did it for PPO: https://github.com/NVIDIA/NeMo-Aligner/blob/main/examples/nlp/gpt/conf/gpt_ppo_actor.yaml#L11

Reply:
SD Trainer raises an error if we add new key to trainer config:

`Traceback (most recent call last):
  File "/opt/nemo-aligner/examples/mm/stable_diffusion/train_sd_draftp.py", line 53, in main
    trainer = MegatronStableDiffusionTrainerBuilder(cfg).create_trainer()
  File "/opt/NeMo/nemo/collections/nlp/parts/megatron_trainer_builder.py", line 135, in create_trainer
    return Trainer(plugins=plugins, strategy=strategy, **self.cfg.trainer, callbacks=callbacks)
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/utilities/argparse.py", line 70, in insert_env_defaults
    return fn(self, **kwargs)
TypeError: Trainer.__init__() got an unexpected keyword argument 'draftp'`

Reply:
you need to call this function: https://github.com/NVIDIA/NeMo-Aligner/blob/main/examples/nlp/gpt/train_gpt_ppo_actor.py#L59

Reply:
to remove that extra field(a bit hacky i know)